cmd/libsnap-confine-private/tool.c

```diff
@@ -76,16 +76,59 @@ void sc_call_snap_update_ns(int snap_update_ns_fd, const char *snap_name,
 {
 	char *snap_name_copy SC_CLEANUP(sc_cleanup_string) = NULL;
 	snap_name_copy = sc_strdup(snap_name);
-	char *argv[] =
-	    { "snap-update-ns", "--from-snap-confine", snap_name_copy, NULL };
-	/* SNAPD_DEBUG=x is replaced by sc_call_snapd_tool_with_apparmor with
-	 * either SNAPD_DEBUG=0 or SNAPD_DEBUG=1, see that function for details. */
+
+	char aa_profile[PATH_MAX] = { 0 };
+	sc_must_snprintf(aa_profile, sizeof aa_profile, "snap-update-ns.%s",
+			 snap_name);
+
+	char *argv[] = {
+		"snap-update-ns",
+		/* This tells snap-update-ns we are calling from snap-confine and locking is in place */
+		"--from-snap-confine",
+		snap_name_copy, NULL
+	};
 	char *envp[] = { "SNAPD_DEBUG=x", NULL };
-	char profile[PATH_MAX] = { 0 };
-	sc_must_snprintf(profile, sizeof profile, "snap-update-ns.%s",
+	sc_call_snapd_tool_with_apparmor(snap_update_ns_fd,
+					 "snap-update-ns", apparmor,
+					 aa_profile, argv, envp);
+}
+
+void sc_call_snap_update_ns_as_user(int snap_update_ns_fd,
+				    const char *snap_name,
+				    struct sc_apparmor *apparmor)
+{
+	char *snap_name_copy SC_CLEANUP(sc_cleanup_string) = NULL;
+	snap_name_copy = sc_strdup(snap_name);
+
+	char aa_profile[PATH_MAX] = { 0 };
+	sc_must_snprintf(aa_profile, sizeof aa_profile, "snap-update-ns.%s",
 			 snap_name);
-	sc_call_snapd_tool_with_apparmor(snap_update_ns_fd, "snap-update-ns",
-					 apparmor, profile, argv, envp);
+
+	const char *xdg_runtime_dir = getenv("XDG_RUNTIME_DIR");
+	char xdg_runtime_dir_env[PATH_MAX];
```

Comment:
Since we're putting `XDG_RUNTIME_DIR=%s` in here, I think it should be PATH_MAX + len(XDG_RUNTIME_DIR) to be on the safe side (I know `sc_must_snprintf` will die and not overflow, nonethelss I think the intention should be made clear when declaring size of the buffer).